### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/Pullback): description of the underlying topological space of a fiber product of schemes

### PR DESCRIPTION
**Pull request by chrisflav**
Let `f : X ⟶ S` and `g : Y ⟶ S` be morphisms of schemes. In this PR we describe the underlying
topological space of `pullback f g`, i.e. the fiber product `X ×[S] Y`. In particular, we show that the points of `X ×[S] Y` correspond bijectively to pairs `(z, p)` of triples `z = (x, y, s)` with `f x = s = g y` and prime ideals `q` of `κ(x) ⊗[κ(s)] κ(y)`.

From the valuative criterion project.

Co-authored-by: Qi Ge
Co-authored-by: Andrew Yang

---

- [x] depends on: #15240
- [x] depends on: #17768 
- [x] depends on: #18255

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by erdOne — APPROVED** [review]
LGTM but I also contributed a significant portion of this.

File: Mathlib/Algebra/Category/Ring/Constructions.lean [hunk 1]
```diff
@@ -103,6 +106,28 @@ def pushoutCoconeIsColimit : Limits.IsColimit (pushoutCocone R A B) :=
...
+instance {R M N : Type*} [Field R] [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N]
+    [Nontrivial M] [Nontrivial N] : Nontrivial (M ⊗[R] N) := by
+  rw [← rank_pos_iff_nontrivial (R := R), rank_tensorProduct]
+  simp [rank_pos]
+
+lemma tensorProduct_nontrivial_of_isField {R M N : Type*} [CommRing R]
+    (h : IsField R) [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N]
+    [Nontrivial M] [Nontrivial N] : Nontrivial (M ⊗[R] N) := by
+  letI : Field R := h.toField
+  infer_instance
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/Algebra/Category/Ring/Constructions.lean]: These should not exist at all now that we have faithfully flat modules, updated.

**Review by erdOne — APPROVED** [review]
LGTM

File: Mathlib/AlgebraicGeometry/PullbackCarrier.lean [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+/-!
+# Underlying topological space of fibre product of schemes
+
+Let `f : X ⟶ S` and `g : Y ⟶ S` be morphisms of schemes. In this file we describe the underlying
+topological space of `pullback f g`, i.e. the fiber product `X ×[S] Y`.
+
+## Main results
+
+- `AlgebraicGeomerty.Scheme.Pullback.carrierEquiv`: The bijective correspondence between the points
```
Comment (kim-em) [inline Mathlib/AlgebraicGeometry/PullbackCarrier.lean:17]:
```suggestion
- `AlgebraicGeometry.Scheme.Pullback.carrierEquiv`: The bijective correspondence between the points
```

File: Mathlib/AlgebraicGeometry/PullbackCarrier.lean [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+  x := x
+  y := y
+  s := g.base y
+  hx := h
+  hy := rfl
+
+variable (T : Triplet f g)
+
+/-- Given `x : X` and `y : Y` such that `f x = s = g y`, this is `κ(x) ⊗[κ(s)] κ(y)`. -/
+def tensor : CommRingCat :=
```
Comment (kim-em) [inline Mathlib/AlgebraicGeometry/PullbackCarrier.lean:69]: I would prefer that arguments of `def`s are given explicitly, not via variables, if they are not too long. It is much more readable. 

(i.e. I usually put the `variable` statement *after* the def that needs it.)

File: Mathlib/AlgebraicGeometry/PullbackCarrier.lean [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+
+namespace AlgebraicGeometry.Scheme.Pullback
+
+universe u
+
+variable {X Y S : Scheme.{u}} (f : X ⟶ S) (g : Y ⟶ S)
+
+/-- A `Triplet` over `f : X ⟶ S` and `g : Y ⟶ S` is a triple of points `x : X`, `y : Y`,
+`s : S` such that `f x = s = f y`. -/
+structure Triplet where
```
Comment (kim-em) [inline Mathlib/AlgebraicGeometry/PullbackCarrier.lean:39]: I'm slightly sad that this doesn't have a more descriptive name, but `Base` is all I can think of, which is no better.

**Comment by kim-em** [comment]
bors d+

**Comment by chrisflav** [comment]
Thanks for the reviews!

bors r+

**Pull request closed without merging**